File: finance_core/services/manual_snapshots.py
from __future__ import annotations

import sqlite3
from typing import Any

from finance_core.services.snapshots import get_latest_snapshot, insert_snapshot


def set_bank_total(conn: sqlite3.Connection, amount: int) -> dict[str, Any]:
    return insert_snapshot(conn, bank_total=amount, memo="set-bank")


def set_securities_total(conn: sqlite3.Connection, amount: int) -> dict[str, Any]:
    return insert_snapshot(conn, securities_total=amount, memo="set-securities")


def set_wallet_total(conn: sqlite3.Connection, amount: int) -> dict[str, Any]:
    latest = get_latest_snapshot(conn)
    description = f"cash-set: {latest['wallet_total']:,}円 -> {amount:,}円"
    conn.execute(
        """
        INSERT INTO wallet_transactions (occurred_on, direction, amount, balance_after, description)
        VALUES (date('now', 'localtime'), 'set', ?, ?, ?)
        """,
        (amount, amount, description),
    )
    return insert_snapshot(conn, wallet_total=amount, memo="cash-set")


def cash_add(conn: sqlite3.Connection, amount: int, memo: str) -> dict[str, Any]:
    if amount <= 0:
        raise ValueError("現金追加額は1円以上で指定してください")
    latest = get_latest_snapshot(conn)
    new_wallet = latest["wallet_total"] + amount
    conn.execute(
        """
        INSERT INTO wallet_transactions (occurred_on, direction, amount, balance_after, description)
        VALUES (date('now', 'localtime'), 'in', ?, ?, ?)
        """,
        (amount, new_wallet, memo),
    )
    return insert_snapshot(conn, wallet_total=new_wallet, memo=f"cash-in: {memo}")


def cash_in(conn: sqlite3.Connection, amount: int, memo: str) -> dict[str, Any]:
    return cash_add(conn, amount, memo)


def cash_out(conn: sqlite3.Connection, amount: int, memo: str) -> dict[str, Any]:
    if amount <= 0:
        raise ValueError("支出額は1円以上で指定してください")
    latest = get_latest_snapshot(conn)
    new_wallet = latest["wallet_total"] - amount
    if new_wallet < 0:
        raise ValueError(f"財布残高が不足しています (現在: {latest['wallet_total']:,}円)")
    conn.execute(
        """
        INSERT INTO wallet_transactions (occurred_on, direction, amount, balance_after, description)
        VALUES (date('now', 'localtime'), 'out', ?, ?, ?)
        """,
        (amount, new_wallet, memo),
    )
    return insert_snapshot(conn, wallet_total=new_wallet, memo=f"cash-out: {memo}")

File: cfor_rs/src/services/manual_snapshots.rs
use crate::error::{FinError, Result};
use crate::models::Snapshot;
use crate::services::snapshots::{get_latest_snapshot, insert_snapshot, SnapshotBuilder};
use rusqlite::Connection;

pub fn set_bank_total(conn: &Connection, amount: i64) -> Result<Snapshot> {
    insert_snapshot(
        conn,
        SnapshotBuilder {
            bank_total: Some(amount),
            memo: Some("set-bank".to_string()),
            ..Default::default()
        },
    )
}

pub fn set_securities_total(conn: &Connection, amount: i64) -> Result<Snapshot> {
    insert_snapshot(
        conn,
        SnapshotBuilder {
            securities_total: Some(amount),
            memo: Some("set-securities".to_string()),
            ..Default::default()
        },
    )
}

pub fn set_wallet_total(conn: &Connection, amount: i64) -> Result<Snapshot> {
    let latest = get_latest_snapshot(conn)?;
    let description = format!(
        "cash-set: {}円 -> {}円",
        crate::models::format_amount(latest.wallet_total),
        crate::models::format_amount(amount)
    );
    conn.execute(
        "INSERT INTO wallet_transactions (occurred_on, direction, amount, balance_after, description)
         VALUES (date('now', 'localtime'), 'set', ?1, ?2, ?3)",
        rusqlite::params![amount, amount, description],
    )?;
    insert_snapshot(
        conn,
        SnapshotBuilder {
            wallet_total: Some(amount),
            memo: Some("cash-set".to_string()),
            ..Default::default()
        },
    )
}

pub fn cash_add(conn: &Connection, amount: i64, memo: &str) -> Result<Snapshot> {
    if amount <= 0 {
        return Err(FinError::invalid("現金追加額は1円以上で指定してください"));
    }
    let latest = get_latest_snapshot(conn)?;
    let new_wallet = latest.wallet_total + amount;
    conn.execute(
        "INSERT INTO wallet_transactions (occurred_on, direction, amount, balance_after, description)
         VALUES (date('now', 'localtime'), 'in', ?1, ?2, ?3)",
        rusqlite::params![amount, new_wallet, memo],
    )?;
    insert_snapshot(
        conn,
        SnapshotBuilder {
            wallet_total: Some(new_wallet),
            memo: Some(format!("cash-in: {}", memo)),
            ..Default::default()
        },
    )
}

pub fn cash_out(conn: &Connection, amount: i64, memo: &str) -> Result<Snapshot> {
    if amount <= 0 {
        return Err(FinError::invalid("支出額は1円以上で指定してください"));
    }
    let latest = get_latest_snapshot(conn)?;
    let new_wallet = latest.wallet_total - amount;
    if new_wallet < 0 {
        return Err(FinError::invalid(format!(
            "財布残高が不足しています (現在: {}円)",
            crate::models::format_amount(latest.wallet_total)
        )));
    }
    conn.execute(
        "INSERT INTO wallet_transactions (occurred_on, direction, amount, balance_after, description)
         VALUES (date('now', 'localtime'), 'out', ?1, ?2, ?3)",
        rusqlite::params![amount, new_wallet, memo],
    )?;
    insert_snapshot(
        conn,
        SnapshotBuilder {
            wallet_total: Some(new_wallet),
            memo: Some(format!("cash-out: {}", memo)),
            ..Default::default()
        },
    )
}
